lldb/tools/lldb-dap/LLDBUtils.cpp

```diff
@@ -235,4 +235,15 @@ std::string GetStringValue(const lldb::SBStructuredData &data) {
   return str;
 }
 
+ScopeSyncMode::ScopeSyncMode(lldb::SBDebugger &debugger)
+    : m_debugger(debugger) {
+  assert(m_debugger.GetAsync() && "Debugger not in asynchronous mode!");
```

Comment:
If I had a initCommand that toggled this, would that trigger this assert? e.g.

```
initCommands: ["script lldb.debugger.SetAsync(False)"]
```

Should this be a warning or error message instead of an assert?

Reply:
Yeah, that would create all kind of havoc, including triggering this assert. I'm not convinced this is the right place to flag that (and show a warning/error to the user). We should either prevent the user from changing that behind our back, or make sure we always reset it before giving control to lldb-dap. I'll remove the assert from this PR.

Reply:
I think it’s okay for scripts to have a level of control over the debugger behavior since they’re pretty helpful for extending the debugger, but I agree that if we have some assumptions about the debugger state it’s good to make sure they’re well known.

Reply:
It's not just okay to allow scripts to temporarily change the Sync vrs. Async state of the debugger, it's pretty close to necessary.  Since we don't have a way from the SB API's to hijack the Process listener, the only way you can write a command that does "step, check something, step again" is to change the debugger mode to Sync.  

We could solve this by forcing everyone to write scripted thread plans for this sort of thing, but that's conceptually more difficult, and I think a lot of people aren't willing to figure that out.

Reply:
BTW, Sync vrs. Async really should be a property of the process not the debugger.  It doesn't so much matter when there's one target/process per debugger, but if you have multiple processes, you can't guarantee that all of them want to be in either sync or async mode at the same time.